provider/oci/environ.go

```diff
@@ -223,42 +379,324 @@ func (e *Environ) Destroy(ctx envcontext.ProviderCallContext) error {
 
 // DestroyController implements environs.Environ.
 func (e *Environ) DestroyController(ctx envcontext.ProviderCallContext, controllerUUID string) error {
-	return errors.NotImplementedf("DestroyController")
+	err := e.Destroy(ctx)
+	if err != nil {
+		logger.Errorf("Failed to destroy environment through controller: %s", errors.Trace(err))
+	}
+	instances, err := e.allControllerManagedInstances(controllerUUID)
+	if err != nil {
+		if err == environs.ErrNoInstances {
+			return nil
+		}
+		return errors.Trace(err)
+	}
+	ids := make([]instance.Id, len(instances))
+	for i, val := range instances {
+		ids[i] = val.Id()
+	}
+
+	err = e.StopInstances(ctx, ids...)
+	if err != nil {
+		return err
+	}
+	logger.Debugf("Cleaning up network resources")
+	return e.cleanupNetworksAndSubnets(controllerUUID, "")
 }
 
 // Provider implements environs.Environ.
 func (e *Environ) Provider() environs.EnvironProvider {
-	return nil
+	return e.p
 }
 
-// StorageProviderTypes implements storage.ProviderRegistry.
-func (e *Environ) StorageProviderTypes() ([]storage.ProviderType, error) {
-	return []storage.ProviderType{}, errors.NotImplementedf("StorageProviderTypes")
-}
+// getCloudInitConfig returns a CloudConfig instance. The default oracle images come
+// bundled with iptables-persistent on Ubuntu and firewalld on CentOS, which maintains
+// a number of iptables firewall rules. We need to at least allow the juju API port for state
+// machines. SSH port is allowed by default on linux images.
+func (e *Environ) getCloudInitConfig(series string, apiPort int) (cloudinit.CloudConfig, error) {
+	// TODO (gsamfira): remove this function when the above mention bug is fixed
+	cloudcfg, err := cloudinit.New(series)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
 
-// StorageProvider implements storage.ProviderRegistry.
-func (e *Environ) StorageProvider(t storage.ProviderType) (storage.Provider, error) {
-	return nil, errors.NotFoundf("storage provider %q", t)
+	if apiPort == 0 {
+		return cloudcfg, nil
+	}
+
+	operatingSystem, err := jujuseries.GetOSFromSeries(series)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	switch operatingSystem {
+	case os.Ubuntu:
+		fwCmd := fmt.Sprintf(
+			"/sbin/iptables -I INPUT -p tcp --dport %d -j ACCEPT", apiPort)
+		cloudcfg.AddRunCmd(fwCmd)
+		cloudcfg.AddScripts("/etc/init.d/netfilter-persistent save")
+	case os.CentOS:
+		fwCmd := fmt.Sprintf("firewall-cmd --zone=public --add-port=%d/tcp --permanent --reload", apiPort)
```

Comment:
I did a quick test of combining ```--reload``` and ```--permanent``` with another firewall mod and got the message:
> Can't use stand-alone options with other options.

I think the reload must be issued by itself.